File: src/test/java/com/mercateo/jsonschema/property/UnwrappedPropertyBuilderTest.java
package com.mercateo.jsonschema.property;

import javaslang.collection.List;
import org.junit.Before;
import org.junit.Test;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static com.mercateo.jsonschema.property.PropertyBuilderTest.getFirstElement;
import static org.assertj.core.api.Assertions.assertThat;

public class UnwrappedPropertyBuilderTest {

    private UnwrappedPropertyBuilder unwrappedPropertyBuilder;

    @Retention(RetentionPolicy.RUNTIME)
    static @interface Unwrap {

    }

    @Before
    public void setUp() throws Exception {
        PropertyBuilder propertyBuilder = new PropertyBuilderDefault(List.of(new FieldCollector(FieldCollectorConfig.builder()
                .build())));
        unwrappedPropertyBuilder = new UnwrappedPropertyBuilder(propertyBuilder, Unwrap.class);
    }

    @Test
    public void singleLevelUnwrap() throws Exception {
        final Property unwrappedProperty = unwrappedPropertyBuilder.from(PropertyHolder.class);

        assertThat(unwrappedProperty.children()).extracting(Property::name).containsExactlyInAnyOrder("foo", "bar");
    }

    @Test
    public void twoLevelUnwrap() throws Exception {
        final Property unwrappedProperty = unwrappedPropertyBuilder.from(SecondLevelPropertyHolder.class);

        assertThat(unwrappedProperty.children()).extracting(Property::name).containsExactlyInAnyOrder("foo", "bar");
    }

    @Test
    public void singleLevelUnwrapGetValue() throws Exception {
        final Property unwrappedProperty = unwrappedPropertyBuilder.from(PropertyHolder.class);

        final PropertyHolder propertyHolder = new PropertyHolder();
        propertyHolder.unwrappedPropertyHolder = new UnwrappedPropertyHolder();
        propertyHolder.unwrappedPropertyHolder.foo = "value1";

        final Property firstElement = getFirstElement(unwrappedProperty.children());

        assertThat(firstElement.getValue(propertyHolder)).isEqualTo("value1");
    }

    static class SecondLevelPropertyHolder {
        @Unwrap
        PropertyHolder propertyHolder;
    }

    static class PropertyHolder {
        @Unwrap
        UnwrappedPropertyHolder unwrappedPropertyHolder;
    }

    static class UnwrappedPropertyHolder {
        String foo;

        String bar;
    }
}

File: src/test/java/com/mercateo/jsonschema/property/UnwrappedPropertyBuilderTest.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.property.UnwrappedPropertyBuilderClasses.*
import org.junit.Before
import org.junit.Test

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.IterableAssert

class UnwrappedPropertyBuilderTest {

    private var unwrappedPropertyBuilder: UnwrappedPropertyBuilder? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        val propertyBuilder = PropertyBuilderDefault(listOf(FieldCollector(FieldCollectorConfig())))
        unwrappedPropertyBuilder = UnwrappedPropertyBuilder(propertyBuilder, Unwrap::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun singleLevelUnwrap() {
        val unwrappedProperty = unwrappedPropertyBuilder!!.from(PropertyHolder::class.java)

        assertThat(unwrappedProperty.children).extracting("name").containsExactlyInAnyOrder("foo", "bar")
    }

    @Test
    @Throws(Exception::class)
    fun twoLevelUnwrap() {
        val unwrappedProperty = unwrappedPropertyBuilder!!.from(SecondLevelPropertyHolder::class.java)

        assertThat(unwrappedProperty.children).extracting("name").containsExactlyInAnyOrder("foo", "bar")
    }

    fun <T> assertThat(actual: Iterable<T>) = IterableAssert(actual)

    @Test
    @Throws(Exception::class)
    fun singleLevelUnwrapGetValue() {
        val unwrappedProperty = unwrappedPropertyBuilder!!.from(PropertyHolder::class.java)

        val propertyHolder = PropertyHolder()
        propertyHolder.unwrappedPropertyHolder = UnwrappedPropertyHolder()
        propertyHolder.unwrappedPropertyHolder!!.foo = "value1"

        val firstElement = unwrappedProperty.children.first()

        assertThat(firstElement.getValue(propertyHolder)).isEqualTo("value1")
    }


}
